unittest/test_costs.cpp

```diff
@@ -0,0 +1,159 @@
+  const boost::shared_ptr<crocoddyl::CostDataAbstract>& data = model->createData(&shared_data);
+
+  // Create the equivalent num diff model and data.
+  crocoddyl::CostNumDiffModel model_num_diff(model);
+  const boost::shared_ptr<crocoddyl::CostDataAbstract>& data_num_diff = model_num_diff.createData(&shared_data);
+
+  // Generating random values for the state and control
+  const Eigen::VectorXd& x = model->get_state()->rand();
+  const Eigen::VectorXd& u = Eigen::VectorXd::Random(model->get_nu());
+
+  // Computing the action derivatives
+  model->calc(data, x, u);
+
+  model_num_diff.calc(data_num_diff, x, u);
+
+  // Checking the partial derivatives against NumDiff
+  BOOST_CHECK(data->cost == data_num_diff->cost);
+}
+
+void test_partial_derivatives_against_numdiff(CostModelTypes::Type cost_type,
+                                              ActivationModelTypes::Type activation_type,
+                                              StateTypes::Type state_multibody_type) {
+  // create the model
+  CostModelFactory factory(cost_type, activation_type, state_multibody_type);
+  const boost::shared_ptr<crocoddyl::CostModelAbstract>& model = factory.get_cost();
+
+  // create the corresponding data object
+  pinocchio::Model& pinocchio_model = model->get_state()->get_pinocchio();
+  pinocchio::Data pinocchio_data(pinocchio_model);
+  crocoddyl::DataCollectorMultibody shared_data(&pinocchio_data);
+  const boost::shared_ptr<crocoddyl::CostDataAbstract>& data = model->createData(&shared_data);
+
+  // Create the equivalent num diff model and data.
+  crocoddyl::CostNumDiffModel model_num_diff(model);
+  const boost::shared_ptr<crocoddyl::CostDataAbstract>& data_num_diff = model_num_diff.createData(&shared_data);
+
+  // Generating random values for the state and control
+  const Eigen::VectorXd& x = model->get_state()->rand();
+  const Eigen::VectorXd& u = Eigen::VectorXd::Random(model->get_nu());
+
+  // Compute all the pinocchio function needed for the models.
+  crocoddyl::updateAllPinocchio(&pinocchio_model, &pinocchio_data, x);
+
+  // set the function that needs to be called at every step of the numdiff
+  std::vector<crocoddyl::ReevaluationFunction> reevals;
+  reevals.push_back(boost::bind(&crocoddyl::updateAllPinocchio, &pinocchio_model, &pinocchio_data, _1));
+  model_num_diff.set_reevals(reevals);
+
+  // Computing the action derivatives
+  model->calcDiff(data, x, u);
+  // Computing the action derivatives via numerical differientiation
+  model_num_diff.calcDiff(data_num_diff, x, u);
+
+  // Checking the partial derivatives against NumDiff
+  double tol = factory.get_num_diff_modifier() * model_num_diff.get_disturbance();
+
+  BOOST_CHECK((data->Lx - data_num_diff->Lx).isMuchSmallerThan(1.0, tol));
+  BOOST_CHECK((data->Lu - data_num_diff->Lu).isMuchSmallerThan(1.0, tol));
+  if (model_num_diff.get_with_gauss_approx()) {
+    // The num diff is not precise enough to be tested here.
```

Comment:
The problem is not the tolerance, we need to do extra work for computing 2nd order derivatives; the same for the action numdiff.

It would be great if you can add this comment.